### ISSUE TITLE
CUDA support in iwyu

### ISSUE THREAD

**Issue by srogatch** [include-what-you-use/include-what-you-use#950]
I don't see in the docs or the search on the internet any evidence of CUDA support in `iwyu`. Can you extend the tool to analyze `.cu` and `.cuh` files?

**Comment by kimgr** [comment]
You could try adding `.cu` as a known source file extension here:
https://github.com/include-what-you-use/include-what-you-use/blob/master/iwyu_path_util.cc#L31

and here:
https://github.com/include-what-you-use/include-what-you-use/blob/master/fix_includes.py#L159

If that seems to work, I'd be happy to take a pull request for it. I don't have any CUDA experience myself, so I don't know how to test it in a reasonable way.

**Comment by srogatch** [comment]
Thanks, @kimgr . Do I understand correctly that the header files such as `*.cuh` do not need a registration and are identified automatically when an `#include` directive is used? How deep does `iwyu` analyze the C++ code? I'm asking because CUDA has some syntax extensions, so to understand whether this could be a problem.

**Comment by kimgr** [comment]
@srogatch I _think_ we only do negative testing to determine what's a header, i.e. `not is-source-file`, so header extensions should not be necessary.

But yes, IWYU sees all `#include` directives. As for depth, it looks at the Clang AST, so depending on how the CUDA extensions manifest there, they might or might not throw IWYU off. If there are any CUDA-specific rules for full-use vs forward-declare use, that would need to be implemented separately, for example.

If I understand CUDA correctly, there's also some kind of host/device divide -- not sure how that looks in code. Are there host-specific and device-specific headers?

**Comment by srogatch** [comment]
Yes, CUDA has member-level granularity for host/device division - a function, a variable, etc. Headers that have device-specific code are normally called `*.cuh` rather than `*.h` and cannot be included in non-CUDA compiled files (normally that means that `*.cuh` files can be included by `*.cu` files, but not `*.cpp` files). That's mostly it for host/device division for headers. I've made the changes you suggested and now analyzing our program with `iwyu`. I'll let you know soon what happened to `*.cu` files.

**Comment by srogatch** [comment]
So at least when run from `cmake`, the `*.cu` files are still ignored after these changes because `cmake` doesn't seem to know of `CMAKE_CUDA_INCLUDE_WHAT_YOU_USE` flag.

**Comment by srogatch** [comment]
I've tried `make -j 18 -k CXX=include-what-you-use` after configuring the project with `cmake`, but the `*.cu` files are still ignored. Perhaps after configuring with `cmake`, `make` just ignores the `CXX` flag or there is a separate flag for CUDA compiler in `make`, but then `iwyu` would need to call `nvcc` instead of `clang` for parsing the CUDA files.

**Comment by kimgr** [comment]
It wouldn't surprise me if `CMAKE_*_INCLUDE_WHAT_YOU_USE` is only supported for languages C and CXX. That mode is maintained by the CMake project, so you might want to raise the question there.

I think CMake-generated builds usually don't honor `CC` and `CXX`, the reasoning is that if you wanted to override, you'd do so at CMake-time. But that won't work for the IWYU scenario, as IWYU doesn't generate object files when used for compiler-testing.

>  but then iwyu would need to call nvcc instead of clang for parsing the CUDA files.

Ah, this is a complication I wasn't aware of -- is the device code handled by `nvcc`? We don't "call" clang per se, we link with its libraries and use them to parse and interrogate the resulting AST and preprocessor state. So if Clang would normally call out to `nvcc` for parsing support, it might do so when driven from IWYU too. But I'm not sure how the bridge between them looks.

**Comment by bgd54** [comment]
Hi!

Is there any update on CUDA support?
I've tried to run iwyu (0.17 with clang 13.0) on a small project with the compile_commands.json that I generated with CMake. Clang has two passes on CUDA files (one for host code and one for device), hence iwyu and other clang tools consider the compile command for `*cu` files as multiple compiler jobs. I manually added `--cuda-host-only` to my compile_commands.json and got reasonable results.
There are some clang specific headers and some differences between the result with `--cuda-host-only` and `--cuda-device-only`, but generally it seems to work.

**Comment by kimgr** [comment]
@bgd54 I don't know of any developments. As far as CUDA, patches are welcome, I don't know enough about the model to reason about how it should be. But let's maybe talk through implementation ideas here first, to make the high-level approach can align with IWYU architecture.
